File: proxy-pattern/src/main/java/com/example/proxy_pattern/service/ReportGeneratorService.java
package com.example.proxy_pattern.service;

import org.springframework.stereotype.Service;

import com.example.proxy_pattern.dto.ReportGeneratedResponseDto;
import com.example.proxy_pattern.entity.Report;

@Service
public class ReportGeneratorService implements ReportGeneratorInterface {
    public ReportGeneratedResponseDto generate(Report report) throws InterruptedException {
        Thread.sleep(5_000);
        return new ReportGeneratedResponseDto(report.getId());
    }
}

File: proxy-pattern/src/services/ReportGeneratorService.ts
import { Report } from '../models/Report';
import { ReportGeneratedResponseDto } from '../dtos/ReportGeneratedResponseDto';
import { ReportGeneratorInterface } from './ReportGeneratorInterface';

/**
 * Service for generating reports.
 * This is a slow operation that takes 5 seconds to complete.
 */
export class ReportGeneratorService implements ReportGeneratorInterface {
  /**
   * Generates a report.
   * @param report The report to generate.
   * @returns A promise that resolves to the generated report data.
   */
  public async generate(report: Report): Promise<ReportGeneratedResponseDto> {
    // Simulate a time-consuming operation
    await new Promise(resolve => setTimeout(resolve, 5000));
    return new ReportGeneratedResponseDto(report.getId());
  }
}